### PR TITLE
GRIDEDIT-712 Fixed bug when connecting DD meshes

### PR DESCRIPTION
**Pull request by BillSenior**
Added several new unit tests.

### CODE REVIEW

File: libs/MeshKernel/include/MeshKernel/ConnectMeshes.hpp
Comment (ahmad-el-sayed) [inline libs/MeshKernel/include/MeshKernel/ConnectMeshes.hpp:1]: All methods, public and private, can be made static.

File: libs/MeshKernel/src/ConnectMeshes.cpp [hunk 1]
```diff
@@ -63,6 +65,9 @@ void meshkernel::ConnectMeshes::AreEdgesAdjacent(const Mesh2D& mesh,
         {
             endNode = mesh.m_edges[edge2].second;
         }
+
+        double absCosPhi = std::abs(NormalizedInnerProductTwoSegments(edge1Start, edge1End, edge2Start, edge2End, mesh.m_projection));
+        areAdjacent = (absCosPhi > 0.9 && absCosPhi <= 1.0);
```
Comment (ahmad-el-sayed) [inline libs/MeshKernel/src/ConnectMeshes.cpp:70]: Can the bounds be also (optional) function parameters?

File: libs/MeshKernelApi/src/MeshKernel.cpp [hunk 2]
```diff
@@ -2290,7 +2290,7 @@ namespace meshkernelapi
         return lastExitCode;
     }
 
-    MKERNEL_API int mkernel_mesh2d_connect_meshes(int meshKernelId, const Mesh2D& mesh2d)
+    MKERNEL_API int mkernel_mesh2d_connect_meshes(int meshKernelId, const Mesh2D& mesh2d, double searchFraction)
```
Comment (ahmad-el-sayed) [inline libs/MeshKernelApi/src/MeshKernel.cpp:2293]: Change in signature of API function, must adapt MKPy, MKNet and GEP.

**Review by ahmad-el-sayed — CHANGES_REQUESTED** [review]

File: libs/MeshKernel/include/MeshKernel/ConnectMeshes.hpp
Comment (BillSenior) [reply libs/MeshKernel/include/MeshKernel/ConnectMeshes.hpp:1]: Do you think that I should even make them free functions?

https://github.com/henrytien/effective-cpp-3rd#item-23-prefer-non-member-non-friend-functions-to-member-functions

File: libs/MeshKernel/src/ConnectMeshes.cpp (hunk 1 above)
Comment (BillSenior) [reply libs/MeshKernel/src/ConnectMeshes.cpp:70]: I can do this but I think that the algorithm is quite fragile and, as can be shown with the separation fraction, any small change results in garbage being generated.

But I will certainly make a named constant

File: libs/MeshKernel/include/MeshKernel/ConnectMeshes.hpp
Comment (BillSenior) [reply libs/MeshKernel/include/MeshKernel/ConnectMeshes.hpp:1]: I retract the question above. I think that this class will not be expanded, as its function is quite specific. Its algorithm may, however be improved at some point.

So, I will make all functions static

**Review by ahmad-el-sayed — APPROVED** [review]

File: libs/MeshKernelApi/src/MeshKernel.cpp (hunk 2 above)
Comment (ahmad-el-sayed) [reply libs/MeshKernelApi/src/MeshKernel.cpp:2293]: MKNet and GEP do not support this functionality yet. Adapting MKPy will be done in this [PR](https://github.com/Deltares/MeshKernelPy/pull/93).

**Review by lucacarniato — APPROVED** [review]

**Pull request merged**